nntrainer/src/util_func.cpp

```diff
@@ -76,27 +76,28 @@ Tensor softmax(Tensor t) {
   divisor.setZero();
 
   for (int k = 0; k < batch; k++) {
-    int index = k * height;
+    int index = k * height * width;
+    float m = -std::numeric_limits<float>::max();
+    // find max
     for (int i = 0; i < height; i++) {
       for (int j = 0; j < width; j++) {
-        dp[index + i] += exp(tp[k * height * width + i * width + j]);
+        if (tp[index + i * width + j] > m)
+          m = tp[index + i * width + j];
       }
     }
-  }
 
-  for (int k = 0; k < batch; ++k) {
-    int index = k * height;
-    for (int i = 1; i < height; ++i) {
-      dp[index] += dp[index + i];
+    // shiftx
+    float sum = 0.0;
+    for (int i = 0; i < height; i++) {
+      for (int j = 0; j < width; j++) {
+        dp[index + width * i + j] = exp(tp[index + i * width + j] - m);
+        sum += dp[index + width * i + j];
+      }
     }
-  }
 
-  for (int k = 0; k < batch; k++) {
-    int index = k * height;
     for (int i = 0; i < height; i++) {
       for (int j = 0; j < width; j++) {
-        int id = k * height * width + i * width + j;
-        rp[id] = exp(tp[id]) / dp[index];
+        rp[index + width * i + j] = dp[index + width * i + j] / sum;
```

Comment:
Q. Is it guaranteed that `sum` is never to be zero?

Reply:
exp(x) is always bigger than 0. So IMO, it is guaranteed.

Reply:
Then, it would be better to add `assert (sum > 0)` before entering for loops. Svace may report this issue as a violation.

Reply:
fixed.